sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala

```diff
@@ -382,18 +382,45 @@ private case class PostgresDialect()
       case Types.ARRAY =>
         val tableName = rsmd.getTableName(columnIdx)
         val columnName = rsmd.getColumnName(columnIdx)
-        val query =
-          s"""
-             |SELECT pg_attribute.attndims
-             |FROM pg_attribute
-             |  JOIN pg_class ON pg_attribute.attrelid = pg_class.oid
-             |  JOIN pg_namespace ON pg_class.relnamespace = pg_namespace.oid
-             |WHERE pg_class.relname = '$tableName' and pg_attribute.attname = '$columnName'
-             |""".stripMargin
+
+        /*
+         Spark does not support different dimensionality per row, therefore we retrieve the
+         dimensionality of any row from Postgres. This might fail later on as Postgres allows
+         different dimensions per row for arrays.
+         */
+        val query = s"SELECT array_ndims($columnName) FROM $tableName LIMIT 1"
+        var arrayDimensionalityResolveNeedsFallback = true
+
         try {
           Using.resource(conn.createStatement()) { stmt =>
             Using.resource(stmt.executeQuery(query)) { rs =>
-              if (rs.next()) metadata.putLong("arrayDimension", rs.getLong(1))
+              if (rs.next()) {
+                metadata.putLong("arrayDimension", rs.getLong(1))
+                arrayDimensionalityResolveNeedsFallback = false
+              }
+            }
+          }
+
+          if (arrayDimensionalityResolveNeedsFallback) {
```

Comment:
we can simply move this into the else-block of the first ` if (rs.next())` without creating `arrayDimensionalityResolveNeedsFallback `

Reply:
I thought it would be cleaner with additional nesting of `Using.resource...`. so I introduced `arrayDimensionalityResolveNeedsFallback`. I have refactored now.